django/utils/text.py

```diff
@@ -64,6 +88,94 @@ def _generator():
     return "".join(_generator())
 
 
+def add_truncation_text(text, truncate=None):
+    if truncate is None:
+        truncate = pgettext(
+            "String to return when truncating text", "%(truncated_text)s…"
+        )
+    if "%(truncated_text)s" in truncate:
+        return truncate % {"truncated_text": text}
+    # The truncation text didn't contain the %(truncated_text)s string
+    # replacement argument so just append it to the text.
+    if text.endswith(truncate):
+        # But don't append the truncation text if the current text already
+        # ends in this.
+        return text
+    return f"{text}{truncate}"
+
+
+class TruncateHTMLParser(HTMLParser):
+    class TruncationCompleted(Exception):
+        pass
+
+    def __init__(self, *, num, truncate, convert_charrefs=True):
+        super().__init__(convert_charrefs=convert_charrefs)
+        self.tags = deque()
+        self.output = ""
+        self.remaining = num
+        self.truncate = truncate
+
+    def handle_startendtag(self, tag, attrs):
+        self.handle_starttag(tag, attrs)
+        if tag not in SELF_CLOSING_TAGS:
+            self.handle_endtag(tag)
+
+    def handle_starttag(self, tag, attrs):
+        self.output += self.get_starttag_text()
+        if tag not in SELF_CLOSING_TAGS:
+            self.tags.appendleft(tag)
+
+    def handle_endtag(self, tag):
+        if tag not in SELF_CLOSING_TAGS:
+            self.output += f"</{tag}>"
+            try:
+                self.tags.remove(tag)
+            except ValueError:
+                pass
+
+    def handle_data(self, data):
+        data, output = self.process(data)
+        data_len = len(data)
+        if self.remaining < data_len:
+            self.remaining = 0
+            self.output += add_truncation_text(output, self.truncate)
+            raise self.TruncationCompleted
+        self.remaining -= data_len
+        self.output += output
+
+    def feed(self, data):
+        try:
+            super().feed(data)
+        except self.TruncationCompleted:
+            self.output += "".join([f"</{tag}>" for tag in self.tags])
+            self.tags.clear()
+            self.reset()
+
+
+class TruncateCharsHTMLParser(TruncateHTMLParser):
+    def process(self, data):
+        output = escape("".join(data[: self.remaining]))
+        return data, output
+
+    def feed(self, data):
+        if self.remaining <= 0:
+            self.output += self.truncate
```

Comment:
Following on from the above comment about zero length should give an empty string, this is the bit that causes that. Further, this seems strange. It's also potentially broken as `self.truncate` could, in theory, contain the `%(truncated_text)s` placeholder -- see the `add_truncation_text()` function.

I suggest that we remove both `TruncateCharsHTMLParser.feed()` and `TruncateWordsHTMLParser.feed()` and add the following to the top of the parent `TruncateHTMLParser.feed()`:

```python
        if self.remaining <= 0:
            return
```

Reply:
I tried to fix the zero length case in https://github.com/django/django/pull/16421/commits/c4f609517b99c5b39b934de0c94c7f1ae0f61b43

What do you think?

I needed to manage both the case when the length is actualy zero and when the length of `replacement` is the equal to or greater than `length`.